#!/bin/bash
# STAR Alignment Pipeline for SARS-CoV-2 and Human Genome
# Compatible with high-performance computing environments.

# Input: Trimmed paired-end FASTQ files generated by Trimmomatic
# Files should follow the naming pattern: *_forward_paired.fastq and *_reverse_paired.fastq

# Define FASTQ file arrays
forward_files=(*_forward_paired.fastq)
reverse_files=(*_reverse_paired.fastq)

# =========================================
# SARS-CoV-2 Alignment
# =========================================
# Set the path to your SARS-CoV-2 STAR genome index directory
genome_dir="STAR_INDEX/sars"

# Run STAR alignment for SARS-CoV-2
for index in "${!forward_files[@]}"; do
    forward_file="${forward_files[index]}"
    reverse_file="${reverse_files[index]}"
    output_prefix=$(basename "${forward_file}" _forward_paired.fastq)_SARSCOV2

    STAR --runThreadN 39 \
         --genomeDir "${genome_dir}" \
         --readFilesIn "${forward_file}" "${reverse_file}" \
         --outFileNamePrefix "STAR/sars_quants/${output_prefix}_" \
         --outSAMtype BAM SortedByCoordinate \
         --twopassMode Basic \
         --outFilterMultimapNmax 1 \
         --quantMode GeneCounts
done

# =========================================
# Human Genome Alignment
# =========================================
# Set the path to your Human STAR genome index directory
genome_dir="STAR_INDEX/human"

# Run STAR alignment for Human genome
for index in "${!forward_files[@]}"; do
    forward_file="${forward_files[index]}"
    reverse_file="${reverse_files[index]}"
    output_prefix=$(basename "${forward_file}" _forward_paired.fastq)_HUMAN

    STAR --runThreadN 39 \
         --genomeDir "${genome_dir}" \
         --readFilesIn "${forward_file}" "${reverse_file}" \
         --outFileNamePrefix "STAR/human_quants/${output_prefix}_" \
         --outSAMtype BAM SortedByCoordinate \
         --twopassMode Basic \
         --outFilterMultimapNmax 1 \
         --quantMode GeneCounts
done